server.go

```diff
@@ -710,10 +731,20 @@ func (s *baseServer) handleNewConn(conn quicConn) {
 		}
 	}
 
+	if clientAddrValidated {
+		if s.numHandshakesValidated.Add(-1) < 0 {
+			panic("server BUG: number of validated handshakes negative")
+		}
+	} else {
+		if s.numHandshakesUnvalidated.Add(-1) < 0 {
+			panic("server BUG: number of unvalidated handshakes negative")
+		}
+	}
+
```

Comment:
I think this block needs to be deferred. There are a couple of returns in the block above and we need to *always* decrement this post a handshake completion.

Reply:
Good catch!

Deferring doesn't quite work, since we want to decrement the counter _before_ adding the connection to the server's accept queue. I refactored the function a bit in 3b6f7aa.